### ISSUE TITLE
Master - slave relations

### ISSUE THREAD

**Issue by hansbogert** [StackExchange/dnscontrol#186]
I like bind's way of making slaves which pick up updates from a master.

How does master-slave setups fit in, in a DNSControl world?

**Comment by tlimoncelli** [comment]
dnscontrol doesn't really have a master-slave configuration.  It is a utility that configures other systems, most of which have some kind of master-slave relationship.

For example, we use dnscontrol to generate BIND zone files, which are SCP'ed to BIND servers that are set up in a redundant fashion.  We also use it to manage zones that are on AWS Route53, which has its own master/slave stuff.

As far as dnscontrol itself... most sites keep dnsconfig.js and creds.json.gpg files on a Git Server.  Any changes are made by doing a "git pull" first, then updating the file, and doing a "git push".  If your Git server is fault-tolerant (master-slave), then you can always access those files and make changes.

Does that help?

**Comment by hansbogert** [comment]
> For example, we use dnscontrol to generate BIND zone files, which are SCP'ed to BIND servers that are set up in a redundant fashion. We also use it to manage zones that are on AWS Route53, which has its own master/slave stuff.

Do you mean that you do not rely on, in my example, Bind's mechanisms to distribute the updates to slaves? Instead you'd make your changes in DNSControl, and then update both a master and a slave? Which would make the distinction between master and slave a moot point.

Also to zoom in on a sentence in the quoted paragraph:
>... that are set up in a redundant fashion.

What do you mean exactly by this? That they simply host the same zone/soa ?

**Comment by tlimoncelli** [comment]
I've used DNSControl with BIND two ways:

1.  DNSControl generates a zonefile which is SCP'ed to the DNS BIND master.  The DNS BIND  slaves get updates from the BIND master by the normal zone transfer mechanism.  This always works for slaves. However if the master dies, DNSControl can not update it (obviously) and the slaves will eventually expire. In an emergency we could manually promote a slave to be master, but that is a very manual process.
2.  DNSControl generates a zonefile which is SCP'ed to each DNS BIND server. Each BIND server is configured as a master.  If one master is down when DNSControl wants to SCP a zone update, it will miss that update. When it comes back up, it will serve old records until the next DNSControl run.

> up in a redundant fashion

When I wrote that I had the 2nd option in mind.

Think of DNSControl as a coworker that is willing to update other servers when asked.  Whether or not those servers are fault tolerant, DNSControl will perform the updates. He doesn't care how they are configured.
